src/executor/current_thread.rs

```diff
@@ -0,0 +1,361 @@
+    /// invocation; any other use will result in a panic.
+    pub fn spawn<F>(future: F)
+    where F: Future<Item = (), Error = ()> + 'static
+    {
+        spawn(future, false).unwrap_or_else(|_| {
+            panic!("cannot call `spawn` unless your thread is already \
+                    in the context of a call to `block_on_all` or \
+                    `block_with_init`")
+        })
+    }
+
+    /// Spawns a daemon, which does *not* block the pending `block_on_all` call.
+    ///
+    /// # Panics
+    ///
+    /// This function can only be invoked within a future given to a `block_*`
+    /// invocation; any other use will result in a panic.
+    pub fn spawn_daemon<F>(future: F)
+    where F: Future<Item = (), Error = ()> + 'static
+    {
+        spawn(future, true).unwrap_or_else(|_| {
+            panic!("cannot call `spawn_daemon` unless your thread is already \
+                    in the context of a call to `block_on_all` or \
+                    `block_with_init`")
+        })
+    }
+
+    /// Cancels *all* spawned tasks and daemons.
+    ///
+    /// # Panics
+    ///
+    /// This function can only be invoked within a future given to a `block_*`
+    /// invocation; any other use will result in a panic.
+    pub fn cancel_all_spawned() {
+        CurrentRunner::with(|runner| runner.cancel_all_spawned())
+            .unwrap_or_else(|()| {
+                panic!("cannot call `cancel_all_spawned` unless your thread is \
+                        already in the context of a call to `block_on_all` or \
+                        `block_with_init`")
+            })
+    }
+}
+
+impl<F> Executor<F> for CurrentThread
+where F: Future<Item = (), Error = ()> + 'static
+{
+    fn execute(&self, future: F) -> Result<(), ExecuteError<F>> {
+        spawn(future, false)
+    }
+}
+
+
+impl<F> Executor<F> for DaemonExecutor
+where F: Future<Item = (), Error = ()> + 'static
+{
+    fn execute(&self, future: F) -> Result<(), ExecuteError<F>> {
+        spawn(future, true)
+    }
+}
+
+fn spawn<F>(future: F, daemon: bool) -> Result<(), ExecuteError<F>>
+where F: Future<Item = (), Error = ()> + 'static,
+{
+    CURRENT.with(|current| {
+        if current.scheduler.get().is_null() {
+            Err(ExecuteError::new(ExecuteErrorKind::Shutdown, future))
+        } else {
+            let spawned = SpawnedFuture {
+                daemon: daemon,
+                inner: Task::new(future),
+            };
+
+            unsafe {
+                (*current.scheduler.get()).push(spawned);
+            }
+
+            Ok(())
+        }
+    })
+}
+
+impl TaskRunner {
+    /// Return a new `TaskRunner`
+    fn new(thread_notify: Arc<ThreadNotify>) -> TaskRunner {
+        TaskRunner {
+            non_daemons: 0,
+            scheduler: scheduler::Scheduler::new(thread_notify),
+        }
+    }
+
+    /// Enter a new `TaskRunner` context
+    fn enter<F, A>(f: F) -> Result<A::Item, A::Error>
```

Comment:
Would appreciate a bit more descriptive comments on this and the next couple top-level functions.